File: android/app/src/main/java/com/openledger/android/data/repository/LedgerRepository.kt
package com.openledger.android.data.repository

import com.openledger.android.data.remote.OpenLedgerApi
import com.openledger.android.data.remote.dto.*
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class LedgerRepository @Inject constructor(
    private val api: OpenLedgerApi,
) {
    suspend fun getAccounts(): Result<List<AccountResponse>> = runCatching {
        val response = api.listAccounts()
        if (!response.isSuccessful) throw Exception("Failed: ${response.code()}")
        response.body()!!
    }

    suspend fun getJournalEntries(): Result<List<JournalEntryResponse>> = runCatching {
        val response = api.listJournalEntries()
        if (!response.isSuccessful) throw Exception("Failed: ${response.code()}")
        response.body()!!
    }

    suspend fun createJournalEntry(entry: JournalEntryCreate): Result<JournalEntryResponse> = runCatching {
        val response = api.createJournalEntry(entry)
        if (!response.isSuccessful) throw Exception("Failed: ${response.code()}")
        response.body()!!
    }

    suspend fun getTransactions(): Result<List<BankTransactionResponse>> = runCatching {
        val response = api.listTransactions()
        if (!response.isSuccessful) throw Exception("Failed: ${response.code()}")
        response.body()!!
    }

    suspend fun generateReport(request: ReportRequest): Result<ReportResponse> = runCatching {
        val response = when (request.reportType) {
            "profit_loss" -> api.profitLoss(request)
            "balance_sheet" -> api.balanceSheet(request)
            "trial_balance" -> api.trialBalance(request)
            else -> throw Exception("Unknown report type: ${request.reportType}")
        }
        if (!response.isSuccessful) throw Exception("Failed: ${response.code()}")
        response.body()!!
    }

    suspend fun getMe(): Result<UserResponse> = runCatching {
        val response = api.getMe()
        if (!response.isSuccessful) throw Exception("Failed: ${response.code()}")
        response.body()!!
    }

    suspend fun getPeriods(): Result<List<PeriodResponse>> = runCatching {
        val response = api.listPeriods()
        if (!response.isSuccessful) throw Exception("Failed: ${response.code()}")
        response.body()!!
    }

    suspend fun getAuditLogs(): Result<List<AuditLogResponse>> = runCatching {
        val response = api.listAuditLogs()
        if (!response.isSuccessful) throw Exception("Failed: ${response.code()}")
        response.body()!!
    }

    suspend fun aiQuery(query: String): Result<AiQueryResponse> = runCatching {
        val response = api.aiQuery(AiQueryRequest(query))
        if (!response.isSuccessful) throw Exception("Failed: ${response.code()}")
        response.body()!!
    }

    suspend fun autoMatch(): Result<List<ReconciliationSuggestion>> = runCatching {
        val response = api.autoMatch()
        if (!response.isSuccessful) throw Exception("Failed: ${response.code()}")
        response.body()!!
    }

    suspend fun getUnmatched(): Result<List<BankTransactionResponse>> = runCatching {
        val response = api.unmatchedItems()
        if (!response.isSuccessful) throw Exception("Failed: ${response.code()}")
        response.body()!!
    }
}

File: ios/OpenLedger/OpenLedger/Data/Repository/LedgerRepository.swift
import Foundation

@Observable
final class LedgerRepository {

    private let api: APIClient

    init(api: APIClient) {
        self.api = api
    }

    func getAccounts() async throws -> [AccountResponse] {
        try await api.request("GET", path: "/api/accounts/")
    }

    func getJournalEntries() async throws -> [JournalEntryResponse] {
        try await api.request("GET", path: "/api/journal-entries/")
    }

    func createJournalEntry(_ entry: JournalEntryCreate) async throws -> JournalEntryResponse {
        try await api.request("POST", path: "/api/journal-entries/", body: entry)
    }

    func getTransactions() async throws -> [BankTransactionResponse] {
        try await api.request("GET", path: "/api/transactions/")
    }

    func generateReport(_ request: ReportRequest) async throws -> ReportResponse {
        let path: String
        switch request.reportType {
        case "profit_loss":
            path = "/api/reports/profit-loss"
        case "balance_sheet":
            path = "/api/reports/balance-sheet"
        case "trial_balance":
            path = "/api/reports/trial-balance"
        default:
            throw APIError.httpError(400, "Unknown report type: \(request.reportType)")
        }
        return try await api.request("POST", path: path, body: request)
    }

    func getMe() async throws -> UserResponse {
        try await api.request("GET", path: "/api/auth/me")
    }

    func getPeriods() async throws -> [PeriodResponse] {
        try await api.request("GET", path: "/api/periods/")
    }

    func getAuditLogs() async throws -> [AuditLogResponse] {
        try await api.request("GET", path: "/api/audit-logs/")
    }

    func aiQuery(_ query: String) async throws -> AiQueryResponse {
        let body = AiQueryRequest(query: query)
        return try await api.request("POST", path: "/api/ai/query", body: body)
    }

    func autoMatch() async throws -> [ReconciliationSuggestion] {
        try await api.request("POST", path: "/api/reconciliation/auto-match")
    }

    func getUnmatched() async throws -> [BankTransactionResponse] {
        try await api.request("GET", path: "/api/reconciliation/unmatched")
    }
}
